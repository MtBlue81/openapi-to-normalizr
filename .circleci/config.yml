version: 2
base_image: &base_image
  image: circleci/node:10.15.1

jobs:
  npm_install:
    docker:
      - <<: *base_image
    steps:
      - checkout
      - restore_cache:
          name: Restore Yarn Package Cache
          keys:
            - yarn-packages-v2-{{ .Branch }}-{{ checksum "yarn.lock" }}
            - yarn-packages-v2-{{ .Branch }}
            - yarn-packages-v2-master
            - yarn-packages-v2-
      - run:
          name: Install Dependencies
          command: yarn install --frozen-lockfile
      - save_cache:
          name: Save Yarn Package Cache
          key: yarn-packages-v2-{{ .Branch }}-{{ checksum "yarn.lock" }}
          paths:
            - node_modules/
  lint:
    docker:
      - <<: *base_image
    steps:
      - checkout
      - restore_cache:
          name: Restore Yarn Package Cache
          keys:
            - yarn-packages-v2-{{ .Branch }}-{{ checksum "yarn.lock" }}
            - yarn-packages-v2-{{ .Branch }}
            - yarn-packages-v2-master
            - yarn-packages-v2-
      - run:
          name: Spec lint
          command: yarn speclint
  test:
    docker:
      - <<: *base_image
    steps:
      - checkout
      - restore_cache:
          name: Restore Yarn Package Cache
          keys:
            - yarn-packages-v2-{{ .Branch }}-{{ checksum "yarn.lock" }}
            - yarn-packages-v2-{{ .Branch }}
            - yarn-packages-v2-master
            - yarn-packages-v2-
      - run:
          name: Test
          command: yarn test
workflows:
  version: 2
  test:
    jobs:
      - npm_install
      - lint:
          requires:
            - npm_install
      - test:
          requires:
            - npm_install
