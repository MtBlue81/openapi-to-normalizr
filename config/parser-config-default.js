const path = require('path');
const templates = {
  model: 'templates/model_template.mustache',
  override: 'templates/override_template.mustache',
  schema: 'templates/schema_template.mustache',
  head: 'templates/head_template.mustache',
  dependency: 'templates/dependency_template.mustache',
  models: 'templates/models_template.mustache',
  actionTypes: 'templates/action_types_template.mustache',
  spec: 'templates/spec_template.mustache',
  oneOf: 'templates/one_of_template.mustache',
};
module.exports = {
  templates: Object.keys(templates).reduce((acc, key) => {
    acc[key] = path.join(__dirname, '../', templates[key]);
    return acc;
  }, {}),
  modelsDir: 'tmp',
  outputDir: 'tmp',
  useFlow: true,
  usePropType: true,
};
