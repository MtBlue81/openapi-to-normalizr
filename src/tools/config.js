class Config {
  constructor(config) {
    this._config = config;
    this.attributeConverter = config.attributeConverter ? config.attributeConverter : (str) => str;
    this.modelsDir = config.modelsDir || 'dist';
  }

  get tags() {
    return this._config.tags;
  }

  get outputPath() {
    return this._config.outputPath;
  }

  get useTypeScript() {
    return this._config.useTypeScript;
  }

  get extension() {
    return this._config.useTypeScript ? 'ts' : 'js';
  }

  formatForModelGenerator() {
    const {
      modelsDir: outputDir,
      templates: templatePath,
      usePropType,
      useFlow,
      useTypeScript,
      attributeConverter,
    } = this._config;
    return {
      outputDir,
      templatePath,
      usePropType,
      useFlow,
      useTypeScript,
      attributeConverter,
      extension: this.extension,
    };
  }

  formatForActionTypesGenerator() {
    const { outputPath, templates: templatePath, useTypeScript } = this._config;
    return {
      templatePath,
      outputPath: outputPath.actions,
      schemasFilePath: outputPath.schemas,
      useTypeScript,
      extension: this.extension,
    };
  }

  formatForSchemaGenerator() {
    const {
      outputPath,
      templates: templatePath,
      modelsDir,
      attributeConverter,
      useTypeScript,
    } = this._config;
    return {
      templatePath,
      outputPath: outputPath.schemas,
      modelsDir,
      attributeConverter,
      useTypeScript,
      extension: this.extension,
    };
  }

  formatForJsSpecGenerator() {
    const { templates: templatePath, outputPath } = this._config;
    return {
      templatePath,
      outputPath: outputPath.jsSpec,
      extension: this.extension,
    };
  }
}

module.exports = Config;
