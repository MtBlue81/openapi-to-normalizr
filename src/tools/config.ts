export default class Config {
  _config: ConfigObject;

  attributeConverter: AttributeConverter;

  modelsDir: string;

  constructor(config: ConfigObject) {
    this._config = config;
    this.attributeConverter = config.attributeConverter ? config.attributeConverter : (str) => str;
    this.modelsDir = config.modelsDir || 'dist';
  }

  get tags() {
    return this._config.tags;
  }

  get outputPath() {
    return this._config.outputPath;
  }

  get useTypeScript() {
    return this._config.useTypeScript;
  }

  get extension() {
    return this._config.useTypeScript ? 'ts' : 'js';
  }

  formatForModelGenerator() {
    const { templates: templatePath, usePropType, useTypeScript } = this._config;
    return {
      outputDir: this.modelsDir,
      templatePath,
      usePropType,
      useTypeScript,
      attributeConverter: this.attributeConverter,
      extension: this.extension,
    } as const;
  }

  formatForActionTypesGenerator() {
    const { outputPath, templates: templatePath, useTypeScript } = this._config;
    const operationIdList: string[] = [];
    return {
      outputPath: outputPath.actions,
      schemasFilePath: outputPath.schemas,
      templatePath,
      operationIdList,
      useTypeScript,
      extension: this.extension,
    } as const;
  }

  formatForSchemaGenerator() {
    const { outputPath, templates: templatePath, useTypeScript } = this._config;
    return {
      templatePath,
      outputPath: outputPath.schemas,
      modelsDir: this.modelsDir,
      attributeConverter: this.attributeConverter,
      useTypeScript,
      extension: this.extension,
    } as const;
  }

  formatForJsSpecGenerator() {
    const { templates: templatePath, outputPath } = this._config;
    return {
      templatePath,
      outputPath: outputPath.jsSpec,
      extension: this.extension,
    } as const;
  }
}
