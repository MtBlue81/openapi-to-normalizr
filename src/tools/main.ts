// @ts-nocheck
import path from 'path';
import Swagger from 'swagger-client';
import _ from 'lodash';
import { mkdirpPromise, getModelDefinitions } from './utils';
import {
  readSpecFilePromise,
  getPreparedSpecFilePaths,
  convertToLocalDefinition,
} from './spec_file_utils';
import ModelGenerator from './model_generator';
import SchemaGenerator from './schema_generator';
import ActionTypesGenerator from './action_types_generator';
import JsSpecGenerator from './js_spec_generator';
import Config from './config';

export default function main(specFiles, c) {
  const config = new Config(c);
  const filePaths = getPreparedSpecFilePaths(specFiles, config.tags);

  Promise.all(
    filePaths.map((file) =>
      readSpecFilePromise(file, {
        dereference: true,
      }),
    ),
  )
    .then((schemas) => {
      const spec = _.merge(...schemas);

      let actionTypesGenerator, modelGenerator, schemaGenerator;
      const [actionsDir, schemasDir, specDir] = ['actions', 'schemas', 'jsSpec'].map((key) =>
        path.dirname(config.outputPath[key]),
      );
      const baseModelsDir = `${config.modelsDir}/base`;
      const prepareDirs = [
        actionsDir,
        schemasDir,
        specDir,
        config.modelsDir,
        baseModelsDir,
      ].map((p) => mkdirpPromise(p));

      return Swagger({
        spec,
      })
        .then(({ spec }) => {
          // refとOpenAPI記法(oneOfなど)解決済みのspecからモデル定義を取得
          const definitions = getModelDefinitions(spec);

          return Promise.all(prepareDirs).then(() => {
            const configForModelGenerator = Object.assign(
              {
                outputBaseDir: baseModelsDir,
                definitions,
              },
              config.formatForModelGenerator(),
            );

            modelGenerator = new ModelGenerator(configForModelGenerator);
            actionTypesGenerator = new ActionTypesGenerator(config.formatForActionTypesGenerator());

            const configForSchemaGenerator = Object.assign(
              {
                modelGenerator,
              },
              config.formatForSchemaGenerator(),
            );
            schemaGenerator = new SchemaGenerator(configForSchemaGenerator);
            walkResponses(spec.paths, [schemaGenerator.parse, actionTypesGenerator.appendId]);
            return Promise.all([schemaGenerator, actionTypesGenerator].map((g) => g.write()));
          });
        })
        .catch((e) => {
          console.error(`Failed: ${e}`);
          throw e;
        });
    })
    .then(() => {
      // no need dereference for js spec file
      return Promise.all(
        filePaths.map((file) =>
          readSpecFilePromise(file, {
            dereference: false,
          }),
        ),
      ).then((schemas) => {
        const spec = convertToLocalDefinition(_.merge(...schemas));
        return new JsSpecGenerator(config.formatForJsSpecGenerator()).write(spec);
      });
    })
    .catch((e) => {
      console.error(`Failed: ${e}`);
      throw e;
    });

  function opId(operation, path, method) {
    return Swagger.helpers.opId(operation, path, method);
  }

  function walkResponses(paths, onResponses = []) {
    _.each(paths, (operations, path) => {
      _.each(operations, (operation, method) => {
        if (!['GET', 'POST', 'PUT', 'PATCH', 'DELETE'].includes(method.toUpperCase())) {
          // use only RESTful methods
          return;
        }
        if (!_.isObject(operation)) {
          console.warn(`not processed. path:${path}, method:${method}`);
          return;
        }
        if (operation.operationId) {
          console.info(
            `no use specified operationId. path:${path}, method:${method}, operationId:${operation.operationId}`,
          );
          delete operation.operationId;
        }
        const response = operation.responses;
        const id = opId(operation, path, method);
        onResponses.forEach((onResponse) =>
          onResponse(id, response, {
            path,
            method,
            operation,
          }),
        );
      });
    });
  }
}
