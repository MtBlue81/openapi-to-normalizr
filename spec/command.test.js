import config from '../config/parser-config-default';
import rimraf from 'rimraf';
import fs from 'fs';
import path from 'path';
import main from '../src/tools/main';

const outputDir = config.modelsDir;
const dir = __dirname;

const walk = (p, fileCallback) => {
  const files = fs.readdirSync(p);
  files.forEach((f) => {
    const fp = path.join(p, f);
    if (fs.statSync(fp).isDirectory()) {
      walk(fp, fileCallback);
    } else {
      fileCallback(fp);
    }
  });
};

const snapshot = (files, customConfigFilePath) => {
  const customConfigOption = customConfigFilePath
    ? require(path.join(dir, customConfigFilePath)) // eslint-disable-line global-require
    : null;
  main(files.map((f) => path.join(dir, f)).join(' '), customConfigOption || config);
  walk(outputDir, (path) => {
    const output = fs.readFileSync(path, 'utf8');
    expect({ path, output }).toMatchSnapshot();
  });
};

describe('schema generator spec', () => {
  beforeEach(() => new Promise((resolve) => rimraf(outputDir, resolve)));

  test('from json schema ref', () => {
    snapshot(['json_schema_ref.yml']);
  });

  test('from one of check', () => {
    snapshot(['one_of.yml']);
  });

  test('from one of other spec file  check', () => {
    snapshot(['one_of_from_other_file.yml']);
  });

  test('from json schema ref TS', () => {
    snapshot(['json_schema_ref.yml'], './config_ts.js');
  });

  test('from one of check TS', () => {
    snapshot(['one_of.yml'], './config_ts.js');
  });

  test('from one of other spec file  check TS', () => {
    snapshot(['one_of_from_other_file.yml'], './config_ts.js');
  });
});
