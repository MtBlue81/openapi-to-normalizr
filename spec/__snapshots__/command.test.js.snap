// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model generator spec from json schema ref 1`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';

export const GENDER_MALE: string = 'male';
export const GENDER_FEMALE: string = 'female';
export const GENDER_OTHER: string = 'other';

const defaultValues = {
  name: undefined,
  gender: undefined,
};

export const schema = new _schema.Entity('Owner', {}, {idAttribute: 'name'});


/**
 * @params ids : Owner's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Owner extends Record(defaultValues) {
  name: string;
  gender: string;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/owner.js",
}
`;

exports[`model generator spec from json schema ref 2`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';

export const KIND_DOG: string = 'Dog';
export const KIND_CAT: string = 'Cat';

const defaultValues = {
  id: undefined,
  kind: undefined,
};

export const schema = new _schema.Entity('Pet', {}, {idAttribute: 'id'});


/**
 * @params ids : Pet's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Pet extends Record(defaultValues) {
  id: number;
  kind: string;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/pet.js",
}
`;

exports[`model generator spec from json schema ref 3`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';
import Owner, { schema as OwnerSchema } from './owner';

export const KIND_DOG: string = 'Dog';
export const KIND_CAT: string = 'Cat';

const defaultValues = {
  id: undefined,
  name: undefined,
  kind: undefined,
  owner: undefined,
};

export const schema = new _schema.Entity('WrappedPet', {}, {idAttribute: 'id'});
schema.define({
  owner: OwnerSchema
});


/**
 * @params ids : WrappedPet's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class WrappedPet extends Record(defaultValues) {
  id: number;
  name: string;
  kind: string;
  owner: Owner;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/wrapped_pet.js",
}
`;

exports[`model generator spec from json schema ref 4`] = `
Object {
  "output": "/**
 * generated from API definition file
 */

import WrappedPet from './wrapped_pet';
import Pet from './pet';
import Owner from './owner';

export {
  WrappedPet,
  Pet,
  Owner,
};
",
  "path": "tmp/index.js",
}
`;

exports[`model generator spec from json schema ref 5`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Owner from './base/owner';
export * from './base/owner';

export default class Owner extends _Owner {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/owner.js",
}
`;

exports[`model generator spec from json schema ref 6`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Pet from './base/pet';
export * from './base/pet';

export default class Pet extends _Pet {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/pet.js",
}
`;

exports[`model generator spec from json schema ref 7`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _WrappedPet from './base/wrapped_pet';
export * from './base/wrapped_pet';

export default class WrappedPet extends _WrappedPet {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/wrapped_pet.js",
}
`;

exports[`model generator spec parse oneOf type 1`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';
import Dog, { schema as DogSchema } from './dog';
import Cat, { schema as CatSchema } from './cat';


const defaultValues = {
  name: undefined,
  pets: undefined,
};

export const schema = new _schema.Entity('Breeder', {}, {idAttribute: 'name'});
const oneOfSchema1 = new _schema.Union({
  dog: DogSchema,
  cat: CatSchema
}, 'kind');
schema.define({
  pets: [
    oneOfSchema1
  ]
});


/**
 * @params ids : Breeder's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Breeder extends Record(defaultValues) {
  name: string;
  pets: (Dog | Cat)[];
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/breeder.js",
}
`;

exports[`model generator spec parse oneOf type 2`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';

export const KIND_CAT: string = 'cat';

const defaultValues = {
  id: undefined,
  kind: undefined,
  name: undefined,
};

export const schema = new _schema.Entity('Cat', {}, {idAttribute: 'id'});


/**
 * @params ids : Cat's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Cat extends Record(defaultValues) {
  id: number;
  kind: string;
  name: string;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/cat.js",
}
`;

exports[`model generator spec parse oneOf type 3`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';

export const KIND_DOG: string = 'dog';

const defaultValues = {
  id: undefined,
  kind: undefined,
  name: undefined,
};

export const schema = new _schema.Entity('Dog', {}, {idAttribute: 'id'});


/**
 * @params ids : Dog's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Dog extends Record(defaultValues) {
  id: number;
  kind: string;
  name: string;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/dog.js",
}
`;

exports[`model generator spec parse oneOf type 4`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';
import Dog, { schema as DogSchema } from './dog';
import Cat, { schema as CatSchema } from './cat';


const defaultValues = {
  name: undefined,
  pet: undefined,
};

export const schema = new _schema.Entity('Owner', {}, {idAttribute: 'name'});
const oneOfSchema1 = new _schema.Union({
  dog: DogSchema,
  cat: CatSchema
}, 'kind');
schema.define({
  pet: oneOfSchema1
});


/**
 * @params ids : Owner's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Owner extends Record(defaultValues) {
  name: string;
  pet: Dog | Cat;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/owner.js",
}
`;

exports[`model generator spec parse oneOf type 5`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Breeder from './base/breeder';
export * from './base/breeder';

export default class Breeder extends _Breeder {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/breeder.js",
}
`;

exports[`model generator spec parse oneOf type 6`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Cat from './base/cat';
export * from './base/cat';

export default class Cat extends _Cat {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/cat.js",
}
`;

exports[`model generator spec parse oneOf type 7`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Dog from './base/dog';
export * from './base/dog';

export default class Dog extends _Dog {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/dog.js",
}
`;

exports[`model generator spec parse oneOf type 8`] = `
Object {
  "output": "/**
 * generated from API definition file
 */

import Owner from './owner';
import Breeder from './breeder';
import Dog from './dog';
import Cat from './cat';

export {
  Owner,
  Breeder,
  Dog,
  Cat,
};
",
  "path": "tmp/index.js",
}
`;

exports[`model generator spec parse oneOf type 9`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Owner from './base/owner';
export * from './base/owner';

export default class Owner extends _Owner {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/owner.js",
}
`;

exports[`schema generator spec from json schema ref 1`] = `
Object {
  "output": "/* eslint-disable no-underscore-dangle */
/**
 * generated from API definition file
 */

import isFunction from 'lodash/isFunction';
import { createAction } from 'redux-actions';
import schemas from '../schemas/sample_schema';
export const GET_PETS__ID_ = 'GET_PETS__ID_';

export function createOpenApiAction(id, payloadCreator = (params) => params, metaCreator) {
  const meta = {openApi: true, id: id.toLowerCase(), schema: schemas[id.toLowerCase()]};
  let _metaCreator = () => meta;
  if (isFunction(metaCreator)) {
    _metaCreator = (...args) => Object.assign(metaCreator(...args), meta);
  }
  return createAction(id, payloadCreator, _metaCreator);
}
",
  "path": "tmp/action_types/sample.js",
}
`;

exports[`schema generator spec from json schema ref 2`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';

export const GENDER_MALE: string = 'male';
export const GENDER_FEMALE: string = 'female';
export const GENDER_OTHER: string = 'other';

const defaultValues = {
  name: undefined,
  gender: undefined,
};

export const schema = new _schema.Entity('Owner', {}, {idAttribute: 'name'});


/**
 * @params ids : Owner's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Owner extends Record(defaultValues) {
  name: string;
  gender: string;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/owner.js",
}
`;

exports[`schema generator spec from json schema ref 3`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';
import Owner, { schema as OwnerSchema } from './owner';

export const KIND_DOG: string = 'Dog';
export const KIND_CAT: string = 'Cat';

const defaultValues = {
  id: undefined,
  name: undefined,
  kind: undefined,
  owner: undefined,
};

export const schema = new _schema.Entity('WrappedPet', {}, {idAttribute: 'id'});
schema.define({
  owner: OwnerSchema
});


/**
 * @params ids : WrappedPet's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class WrappedPet extends Record(defaultValues) {
  id: number;
  name: string;
  kind: string;
  owner: Owner;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/wrapped_pet.js",
}
`;

exports[`schema generator spec from json schema ref 4`] = `
Object {
  "output": "/**
 * generated from API definition file
 */

import WrappedPet from './wrapped_pet';
import Owner from './owner';

export {
  WrappedPet,
  Owner,
};
",
  "path": "tmp/index.js",
}
`;

exports[`schema generator spec from json schema ref 5`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Owner from './base/owner';
export * from './base/owner';

export default class Owner extends _Owner {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/owner.js",
}
`;

exports[`schema generator spec from json schema ref 6`] = `
Object {
  "output": "/* eslint-disable quotes,comma-dangle */
/**
 * generated from API definition file
 */

export default {
  \\"openapi\\": \\"3.0.0\\",
  \\"info\\": {
    \\"title\\": \\"only components\\",
    \\"version\\": \\"1.0.0\\",
    \\"contact\\": {
      \\"name\\": \\"API Support\\",
      \\"url\\": \\"http://www.example.com/support\\",
      \\"email\\": \\"support@example.com\\"
    }
  },
  \\"tags\\": [
    {
      \\"name\\": \\"default\\",
      \\"description\\": \\"default API\\"
    },
    {
      \\"name\\": \\"default\\",
      \\"description\\": \\"default API\\"
    }
  ],
  \\"paths\\": {
    \\"/pets/{id}\\": {
      \\"get\\": {
        \\"tags\\": [
          \\"pet\\"
        ],
        \\"operationId\\": \\"getPet\\",
        \\"description\\": \\"get pet\\\\n\\",
        \\"parameters\\": [
          {
            \\"description\\": \\"pet id\\",
            \\"name\\": \\"id\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"type\\": \\"integer\\",
              \\"format\\": \\"int64\\"
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"$ref\\": \\"#/components/responses/PetResponse\\"
          }
        }
      }
    }
  },
  \\"components\\": {
    \\"responses\\": {
      \\"PetResponse\\": {
        \\"description\\": \\"a pet response\\",
        \\"content\\": {
          \\"application/json\\": {
            \\"schema\\": {
              \\"$ref\\": \\"#/components/schemas/WrappedPet\\"
            }
          }
        }
      }
    },
    \\"schemas\\": {
      \\"WrappedPet\\": {
        \\"description\\": \\"Wrapped pet\\",
        \\"type\\": \\"object\\",
        \\"required\\": [
          \\"id\\",
          \\"name\\",
          \\"kind\\",
          \\"owner\\"
        ],
        \\"properties\\": {
          \\"id\\": {
            \\"$ref\\": \\"./only_components.yml#/components/schemas/Pet/properties/id\\"
          },
          \\"name\\": {
            \\"type\\": \\"string\\"
          },
          \\"kind\\": {
            \\"$ref\\": \\"#/components/schemas/Pet/properties/kind\\"
          },
          \\"owner\\": {
            \\"$ref\\": \\"#/components/schemas/Owner\\"
          }
        }
      },
      \\"Pet\\": {
        \\"description\\": \\"Pet\\",
        \\"type\\": \\"object\\",
        \\"required\\": [
          \\"id\\",
          \\"name\\"
        ],
        \\"properties\\": {
          \\"id\\": {
            \\"description\\": \\"pet id\\",
            \\"type\\": \\"integer\\",
            \\"format\\": \\"int64\\"
          },
          \\"kind\\": {
            \\"description\\": \\"kind of pet\\",
            \\"type\\": \\"string\\",
            \\"enum\\": [
              \\"Dog\\",
              \\"Cat\\"
            ]
          }
        }
      },
      \\"Owner\\": {
        \\"type\\": \\"object\\",
        \\"x-id-attribute\\": \\"name\\",
        \\"properties\\": {
          \\"name\\": {
            \\"description\\": \\"owner's name\\",
            \\"type\\": \\"string\\"
          },
          \\"gender\\": {
            \\"description\\": \\"owner's gender\\",
            \\"type\\": \\"string\\",
            \\"enum\\": [
              \\"male\\",
              \\"female\\",
              \\"other\\"
            ]
          }
        }
      }
    }
  }
};
",
  "path": "tmp/sample_api.js",
}
`;

exports[`schema generator spec from json schema ref 7`] = `
Object {
  "output": "/* eslint-disable */
/**
 * generated from API definition file
 */

import { schema as WrappedPetSchema } from './../wrapped_pet';

export default {
  get_pets__id_: {
    200: WrappedPetSchema
  }
};
",
  "path": "tmp/schemas/sample_schema.js",
}
`;

exports[`schema generator spec from json schema ref 8`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _WrappedPet from './base/wrapped_pet';
export * from './base/wrapped_pet';

export default class WrappedPet extends _WrappedPet {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/wrapped_pet.js",
}
`;
